libraries/AC_Avoidance/AP_OADijkstra.cpp

```diff
@@ -0,0 +1,550 @@
+                        _short_path_data[item_node_idx].distance_cm = dist_to_item_via_current_node;
+                        _short_path_data[item_node_idx].distance_from_idx = curr_node_idx;
+                    }
+                }
+            }
+        }
+    }
+}
+
+// find a node's index into _short_path_data array from it's id (i.e. id type and id number)
+// returns true if successful and node_idx is updated
+bool AP_OADijkstra::find_node_from_id(const AP_OAVisGraph::OAItemID &id, node_index &node_idx) const
+{
+    // source node is always the first node
+    if ((id.id_type == AP_OAVisGraph::OATYPE_SOURCE) && (_short_path_data_numpoints > 0)) {
+        node_idx = 0;
+        return true;
+    }
+
+    // destination is always the 2nd node
+    if ((id.id_type == AP_OAVisGraph::OATYPE_DESTINATION) && (_short_path_data_numpoints > 1)) {
+        node_idx = 1;
+        return true;
+    }
+
+    // must be a fence node which start from 3rd node
+    if ((id.id_type == AP_OAVisGraph::OATYPE_FENCE_POINT) && (_short_path_data_numpoints > id.id_num + 2)) {
+        node_idx = id.id_num + 2;
+        return true;
+    }
+
+    // could not find node
+    return false;
+}
+
+// find index of node with lowest tentative distance (ignore visited nodes)
+// returns true if successful and node_idx argument is updated
+bool AP_OADijkstra::find_closest_node_idx(node_index &node_idx) const
+{
+    uint16_t lowest_idx = 0;
+    float lowest_dist = 0.0f;
+    bool lowest_found = false;
+
+    // scan through all nodes looking for closest
+    for (uint16_t i=0; i<_short_path_data_numpoints; i++) {
+        const ShortPathNode &node = _short_path_data[i];
+        if (!node.visited && (!lowest_found || node.distance_cm < lowest_dist)) {
+            lowest_found = true;
+            lowest_idx = i;
+            lowest_dist = node.distance_cm;
+        }
+    }
+
+    if (lowest_found) {
+        node_idx = lowest_idx;
+    }
+    return lowest_found;
+}
+
+// calculate shortest path from origin to destination
+// returns true on success
+// requires create_polygon_fence_with_margin and create_polygon_fence_visgraph to have been run
+// resulting path is stored in _shortest_path array as vector offsets from EKF origin
+bool AP_OADijkstra::calc_shortest_path(const Location &origin, const Location &destination)
+{
+    // convert origin and destination to offsets from EKF origin
+    Vector2f origin_NE, destination_NE;
+    if (!origin.get_vector_xy_from_origin_NE(origin_NE) || !destination.get_vector_xy_from_origin_NE(destination_NE)) {
+        return false;
+    }
+
+    // create origin and destination visgraphs of polygon points
+    update_visgraph(_source_visgraph, {AP_OAVisGraph::OATYPE_SOURCE, 0}, origin_NE, true, destination_NE);
+    update_visgraph(_destination_visgraph, {AP_OAVisGraph::OATYPE_DESTINATION, 0}, destination_NE);
+
+    // clear short_path_data array
+    _short_path_data_numpoints = 0;
```

Comment:
This is a very verbose way to set the first and second points with the total number of points set to 2.

Reply:
sure, shortened to 3 lines..